### PR TITLE
Fix complex grain

### PR DESCRIPTION
**Pull request by mickenordin**
### What does this PR do?
If a grain is a list of complex objects, grains.py is unable to compare the currently set grain to the potentially new one. Instead it fails with TypeError: unhashable type: 'dict' (might be something else aswell depending on what you are trying to set as a grain). This PR will make it so that a list grain that contains complex objects can be compared via the function make_hashable.    
### What issues does this PR fix or reference?
issue #39875

### Tests written?

Yes, in tests/unit/states/test_grains.py there is a test called "test_make_hashable"

### Commits signed with GPG?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by waynew** [comment]
Hi! Thanks for the PR!

I think you either meant to make this PR against a different branch or you want to rebase your changes against develop. If you need any pointers I'm happy to help!

**Comment by mickenordin** [comment]
Hi, you are probably right. I started to make a clean PR from my fork to
development, but then I read your development guidelines and tried to
follow the instructions there...

This is a simple fix in a file that has been untuched for a long time. What
is the best way to fix this PR?

Den fre 26 apr. 2019 16:22Wayne Werner <notifications@github.com> skrev:

**Comment by waynew** [comment]
@mickenordin If you've just tried to merge into the wrong branch you can go to the top right and click "edit" and then change what branch you're trying to merge into (probably 2018.3 is my guess).

If that's not right, let me know where it belongs, and where you branched from, and I can help you rebase or cherry pick your changes.

**Comment by mickenordin** [comment]
@waynew thank you very much, now it looks cleaner I think :)

File: salt/states/grains.py [hunk 1]
```diff
@@ -44,6 +44,13 @@ def exists(name, delimiter=DEFAULT_TARGET_DELIM):
         ret['comment'] = 'Grain does not exist'
     return ret
 
+def flatten(li, flattened = list()):
+    for subli in li:
+        if type(subli) == list:
+            flatten(subli, flattened)
+        else:
+            flattened.append(frozenset(subli))
+    return set(flattened)
```
Comment (waynew) [inline salt/states/grains.py]: Would it be better to just use `set` here? The default arg *definitely* should be changed, though.

I'm thinking something like:

```
def flatten(li, flattened=None):
    flattened = flattened or set()
    for subli in li:
        if type(subli) == list:
            flattened.update(flatten(subli))
        else:
            flattened.add(frozenset(subli))
    return flattened
```

I'm not entirely certain that's the best approach either, though.

**Review by waynew — CHANGES_REQUESTED** [review]
Thanks for the PR! Could you update the function definition, as well as add some tests?

`tests/unit/states/test_grains.py` would be a good place to add (or update) tests for this.

File: salt/states/grains.py (hunk 1 above)
Comment (mickenordin) [reply salt/states/grains.py]: Great, I will fix those things :)

**Comment by mickenordin** [comment]
@waynew thank you for you patiens and your help. I have now fixed all lint errors, added a test that uses the function and added comments and over all tried to make the code more clear and precise.

**Review by waynew — APPROVED** [review]

**Comment by mickenordin** [comment]
What is the next step for this PR? Is there a process for merging or is more review needed?

**Comment by waynew** [comment]
@mickenordin shouldn't be - just keep getting changes before we can merge it in 😑 I'll keep my eye out today

**Comment by waynew** [comment]
🎉

**Pull request merged**